docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,1259 @@
+
+In order to ensure that the pod evictions due to VPA recommendations do not increase the risk of a quorum loss, appropriate thresholds should be chosen for VPA to ensure that scaling is not very frequent.
+
+## Handling Outages
+
+### Node failures
+
+It is possible that node(s) hosting the control plane component are no longer available/reachable. Some of the reasons could be crashing of a node, kubelet running on the node is unable to renew its lease, network partition etc. The topology of control plane components and recovery mechanisms will determine the duration of the downtime that will result when a node is no longer reachable/available.
+
+#### Impact of Node failure
+
+It is possible that the nodes hosting shoot control planes either crash or are unable to communicate to its Kube Apiserver to renew their lease.
+
+**Case #1**
+
+HA Type: `single-zone`
+
+For control plane components having multiple replicas, each replica will be provisioned on a different node (one per node) as per the scheduling constraints.
+
+Since there are lesser than the desired replicas for shoot control plane pods, `kube-scheduler` will attempt to look for another node while respecting pod scheduling constraints. If a node satisfying scheduling constraints is found then it will be chosen to schedule control plane pods. If there are no nodes that satisfy the scheduling constraints, then it must wait for `Cluster-Autoscaler` to scale the node group and then for the `Machine-Controller-Manager` to provision a new node in the scaled node group. In the event `Machine-Controller-Manager` is unable to create a new machine, then the replica that was evicted from the failed node, will be stuck in `Pending` state.
+
+_Impact on etcd_
+
+Assuming that default etcd cluster size of 3 members, unavailability of one node is tolerated. If more than 1 node hosting the control plane components goes down or is unavailable, then etcd cluster will lose quorum till new nodes are provisioned.
+
+**Case #2**
+
+HA Type: `multi-zone`
+
+For control plane components having multiple replicas, each replica will be spread across zones as per the scheduling constraints.
+
+_etcd_
+
+`kube-scheduler` will attempt to look for another node in the same zone since the pod scheduling constraints will prevent it from scheduling the pod onto another zone. If a node is found where there are no control plane components deployed, then it will choose that node to schedule the control plane pods. If there are no nodes that satisfy the scheduling constraints, then it must wait for `Machine-Controller-Manager` to provision a new node. Reference to `PVC` will also prevent an `etcd` member from getting scheduled in another zone since persistent volumes are not shared across availability zones.
+
+_Kube Apiserver and Gardener Resource Manager_
```

Comment:
Why would GRM use `preferredDuringSchedulingIgnoredDuringExecution`? It should use `requiredDuringSchedulingIgnoredDuringExecution`. This component doesn't need to run more than 2 or 3 replicas (why does it even need 3 replicas today?).

Reply:
>This component doesn't need to run more than 2 or 3 replicas (why does it even need 3 replicas today?).

We had a similar question, where we were debating on how many replicas should be sufficient to have a failure tolerance of 1 AZ. We concluded that for many of the components it should be at least 2 (and not necessarily 3). GRM today has 3 replicas. Tim clarified this question with Rafael and he was pointed to https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/ which was the influencer. However we can of course question the current state.

Reply:
affinity rule changes are in